### ISSUE TITLE
Narrowing variables that has been captured by closures

### ISSUE THREAD

**Issue by heshanpadmasiri** [ballerina-platform/ballerina-spec#1263]
Consider the fallowing,
```typescript
import ballerina/io;
type F function (int) returns int;
public function main() {
    int val = 10;
    F f = function(int a) returns int {
        val += a;
        return 5;
    };
    match val {
        10 => {
            int res = f(2) + val;
            io:println(res); // @output 17
        }
    }
}
```
Variable `val` has been captured before we narrow the variable in the match statement. Then calling `f` mutates the narrowed value without any direct reference to the value, which should undo the narrowing. Note this is somewhat different from #1043 since we are capturing before narrowing, not after.

**Comment by jclark** [comment]
I think we can make the example simpler.

**Comment by jclark** [comment]
What's the relationship to #1043?

**Comment by heshanpadmasiri** [comment]
Closed in favor of #1043
